src/Products/PluginIndexes/CompositeIndex/CompositeIndex.py

```diff
@@ -0,0 +1,599 @@
+            if rec.keys and operator == 'and':
+                continue
+
+            # continue if no keys in query were set
+            if rec.keys is None:
+                continue
+
+            c_records.append((c.id, rec))
+
+        # return if less than MIN_COMPONENTS query attributes were catched
+        if len(c_records) < MIN_COMPONENTS:
+            return query
+
+        kw_list = []
+        for c_id, rec in c_records:
+            kw = rec.keys
+            if not kw:
+                continue
+            if isinstance(kw, list):
+                kw = tuple(kw)
+            elif not isinstance(kw, tuple):
+                kw = (kw,)
+            kw = tuple([(c_id, k) for k in kw])
+            kw_list.append(kw)
+
+        # permute keyword list
+        records = tuple(product(*kw_list))
+
+        # substitude matching query attributes as composite index
+        cquery.update({self.id: {'query': records}})
+
+        # delete original matching query attributes from query
+        for c_id, rec in c_records:
+            if c_id in cquery:
+                del cquery[c_id]
+
+        #LOG.debug('%s: query build from %r' % (self.__class__.__name__,
+        #                              [(c_id, rec.keys, rec.get('operator'))
+        #                                   for c_id, rec in c_records]))
+
+        return cquery
+
+    def addComponent(self, c_id, c_meta_type, c_attributes):
+        # Add a component object by 'c_id'.
+        if c_id in self._components:
+            raise KeyError('A component with this '
+                           'name already exists: %s' % c_id)
+
+        self._components[c_id] = Component(c_id,
+                                           c_meta_type,
+                                           c_attributes)
+        self.clear()
+
+    def delComponent(self, c_id):
+        # Delete the component object specified by 'c_id'.
+        if c_id not in self._components:
+            raise KeyError('no such Component:  %s' % c_id)
+
+        del self._components[c_id]
+
+        self.clear()
+
+    def saveComponents(self, components):
+        # Change the component object specified by 'c_id'.
+        for c in components:
+            self.delComponent(c.old_id)
+            self.addComponent(c.id, c.meta_type, c.attributes)
+
+    def manage_addComponent(self, c_id, c_meta_type, c_attributes, URL1,
+                            REQUEST=None, RESPONSE=None):
+        """ add a new component """
+        if len(c_id) == 0:
+            raise RuntimeError('Length of component ID too short')
+        if len(c_meta_type) == 0:
+            raise RuntimeError('No component type set')
+
+        self.addComponent(c_id, c_meta_type, c_attributes)
+
+        if RESPONSE:
+            RESPONSE.redirect(URL1 + '/manage_main?'
+                              'manage_tabs_message=Component%20added')
+
+    def manage_delComponents(self, del_ids=[], URL1=None,
```

Comment:
I prefer using an empty immutable tuple, instead of a mutable list in function signatures. It's too easy to get bitten by `del_ids` being a mutable list defined at class scope later on.